cpp/src/plasma/client.cc

```diff
@@ -661,6 +661,10 @@ Status PlasmaClient::Impl::Release(const ObjectID& object_id) {
     auto iter = gpu_object_map.find(object_id);
     ARROW_CHECK(iter != gpu_object_map.end());
     if (--iter->second->client_count == 0) {
+      // Some user may not release the share point of CudaBuffer promptly,
+      // so it is better to call CudaBuffer::Close manually.
+      ARROW_CHECK_OK(iter->second->ptr->Close());
```

Comment:
Since the error is not bubbled up, does it make sense to keep the Close() protected and just invoke it in the destructor?

Reply:
I meant that the the destructor would call Close() and would be automatically reclaimed when the shared_ptr go out of scope.

Reply:
Thank you for your advice. It is modified.

Reply:
Hum, I think you uncovered a bigger issue. `second`s destructor doesn't get called either, thus the shared_ptr handle will not decrease the ref-counter.

Would you know why `static std::unordered_map<ObjectID, GpuProcessHandle*> gpu_object_map;` use a pointer to the GpuProcessHandle and not the object itself?

Seems like the real fix would be to avoid the usage of the pointer (or wrap in a unique_ptr).

Reply:
The following code will cause error.

thread 1:
{
    std::shared_ptr<Buffer> buff;    
    plasma_client1.Create(object_id1, size, nullptr, 0, &buff, 1);
    plasma_client1.Seal(object_id);
    // not to set buff nullptr
    plasma_client1.Release(object_id);
    plasma_client1.Delete(object_id);
    // ... do someting else or not to do anything
}
// let buff auto release here.


thread 2:
{
    std::shared_ptr<Buffer> buff;    
    plasma_client2.Create(object_id2, size, nullptr, 0, &buff, 1);
    // If the address allocated by the server is just the object_id1 released, error occur!
}

Reply:
That's what I'm saying, you shouldn't call the Buffer explicit destructor. The issue is that GpuProcessHandle destructor doesn't get called with the delete. Thus, the shared_ptr destructor doesn't get called, thus the ref-count doesn't get decremented, and when it (if) reaches zero, the Buffer destructor doesn't get called.

Reply:
You effectively uncovered a memory leak.

Reply:
I am so sorry that I just get your point now.

>> Would you know why static std::unordered_map<ObjectID, GpuProcessHandle*> gpu_object_map; use a pointer to the GpuProcessHandle and not the object itself?
>> Seems like the real fix would be to avoid the usage of the pointer (or wrap in a unique_ptr).
I will try to fix it later.
Thank you.